compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConvertToDma.cpp

```diff
@@ -302,25 +303,84 @@ LogicalResult rewriteAsDma(PackOrUnpackOp op, IRRewriter &rewriter) {
+           "to an iree.linalg_ext dialect pack/unpack operation";
+    return failure();
+  }
+  // Setting up the 'identity' pack/unpack:
+  ArrayRef<int64_t> innerDimsPos{};
+  ArrayRef<OpFoldResult> innerTiles{};
+
+  Value src = copyOp.getOperand(0);
+  Value dst = copyOp.getOperand(1);
+
+  // MemRefTypes with no memory space attribute return 0 here, so this is safe.
+  uint32_t srcMemspace = cast<MemRefType>(src.getType()).getMemorySpaceAsInt();
+  uint32_t dstMemspace = cast<MemRefType>(dst.getType()).getMemorySpaceAsInt();
+  // Memory space 0 : L3
+  // Memory space 1 : L2
+  // Memory space 2 : L1
+  // So the copy is towards the core if the source memory space is less than
+  // the destination memory space. We'll convert copies towards the core to
+  // packs and all other copies to unpacks, just so that we don't ever have have
+  // pack ops which move data away from the core -- while valid this will be
+  // potentially confusing IR as packs usually go towards the core, and it might
+  // break implicit developer assumptions if packs are used to move data away
+  // from the core.
+  const bool towardsCore = srcMemspace <= dstMemspace;
+
+  rewriter.setInsertionPoint(copyOp);
+  if (towardsCore) {
+    rewriter.replaceOpWithNewOp<IREE::LinalgExt::PackOp>(
+        copyOp, src, dst, innerDimsPos, innerTiles);
+  } else {
+    rewriter.replaceOpWithNewOp<IREE::LinalgExt::UnPackOp>(
+        copyOp, src, dst, innerDimsPos, innerTiles);
+  }
+
+  return success();
+}
+
 };  // namespace
 
-class AMDAIEPackToDmaPass
-    : public impl::AMDAIEPackToDmaBase<AMDAIEPackToDmaPass> {
+class AMDAIEConvertToDmaPass
+    : public impl::AMDAIEConvertToDmaBase<AMDAIEConvertToDmaPass> {
  public:
   void getDependentDialects(DialectRegistry &registry) const override {
     registry.insert<tensor::TensorDialect, linalg::LinalgDialect,
                     IREE::LinalgExt::IREELinalgExtDialect, AMDAIEDialect>();
   }
 
-  AMDAIEPackToDmaPass() = default;
-  AMDAIEPackToDmaPass(const AMDAIEPackToDmaPass &pass){};
+  AMDAIEConvertToDmaPass() = default;
+  AMDAIEConvertToDmaPass(const AMDAIEConvertToDmaPass &pass){};
   void runOnOperation() override;
 };
 
-void AMDAIEPackToDmaPass::runOnOperation() {
+void AMDAIEConvertToDmaPass::runOnOperation() {
   MLIRContext *context = &getContext();
   IRRewriter rewriter(context);
 
+  // Convert all linalg.copy to iree_linalg_ext.pack/unpack ops. We then
+  // bootstrap the work done for lowering the pack/unpack op to dmas as the next
+  // step. This is easy to implement, but not the most direct lowering, so
+  // we might want to revisit this.
+  WalkResult convertCopiesWalkResult =
+      getOperation()->walk([&rewriter](linalg::CopyOp copyOp) {
+        if (failed(copyToPack(rewriter, copyOp)))
+          return WalkResult::interrupt();
+        return WalkResult::advance();
+      });
+  if (convertCopiesWalkResult.wasInterrupted()) return signalPassFailure();
```

Comment:
It's good to emit the error message.

Reply:
It's already emitted. There is no additional information to add, all failures from copyToPack already emit